src/Compiler/Microsoft.CodeAnalysis.Razor/src/DefaultTypeNameFeature.cs

```diff
@@ -22,24 +22,45 @@ public override IReadOnlyList<string> ParseTypeParameters(string typeName)
         }
 
         var parsed = SyntaxFactory.ParseTypeName(typeName);
+
         if (parsed is IdentifierNameSyntax identifier)
         {
             return Array.Empty<string>();
         }
-        else if (parsed is ArrayTypeSyntax array)
+
+        if (TryParseCore(parsed) is { } list)
         {
-            return new[] { array.ElementType.ToString() };
+            return list;
         }
-        else if (parsed is TupleTypeSyntax tuple)
+
+        return parsed.DescendantNodesAndSelf()
+            .OfType<TypeArgumentListSyntax>()
+            .SelectMany(arg => arg.Arguments)
+            .SelectMany(ParseCore).ToArray();
+
+        static IReadOnlyList<string> TryParseCore(TypeSyntax parsed)
         {
-            return tuple.Elements.Select(a => a.ToString()).ToList();
+            if (parsed is ArrayTypeSyntax array)
+            {
+                return ParseCore(array.ElementType);
+            }
+
+            if (parsed is TupleTypeSyntax tuple)
+            {
+                return tuple.Elements.SelectMany(a => ParseCore(a.Type)).ToList();
```

Comment:
Nit: Can we change the return of this method to `ImmutableArray` and make this `SelectAsArray`?

Reply:
Makes sense, thanks. Changed the code to use immutable arrays. Let me know if you think it could be improved even more.